src/scalib_ext/scalib/src/sasca/factor_graph.rs

```diff
@@ -322,18 +324,91 @@ impl FactorGraph {
                         ));
                     }
                 }
-                FactorKind::GenFactor { .. } => {
-                    // We cannot verify anything, except if we were given the
-                    // value of the generalized factors as argument of this
-                    // function (then we could check that the associated
-                    // probability is non-zero).
-                    // Let us not do any verification for know (code block
-                    // intentionally left empty).
+                FactorKind::GenFactor { id, operands } => {
+                    let ops: Vec<&PublicValue> = operands
+                        .iter()
+                        .map(|op| match op {
+                            GenFactorOperand::Var(idx, ..) => &var_assignments[*idx],
+                            GenFactorOperand::Pub(idx) => &public_values[*idx],
+                        })
+                        .collect();
+                    let mut ops_iter = ops.iter();
+                    let nmulti = loop {
+                        if let Some(PublicValue::Multi(x)) = ops_iter.next() {
+                            break x.len();
+                        }
+                    };
+                    let mut indices: Vec<Vec<ClassVal>> = vec![vec![0; ops.iter().len()]; nmulti];
+                    for i in 0..nmulti {
+                        for (j, pv) in ops.iter().enumerate() {
+                            indices[i][j] = match *pv {
+                                PublicValue::Single(x) => *x,
+                                PublicValue::Multi(xvec) => xvec[i],
+                            };
+                        }
+                    }
+                    match &gen_factors[*id] {
+                        super::GenFactor::Single(gf) => match gf {
```

Comment:
There is a lot of duplication in the Single & Multi cases. To reduce that, one solution is to make an outer loop `for i in 0..nmulti`, in which you first extract the factor `gf`/`gf[i]`, then do the check according to dense/sparsefunctional.